### PR TITLE
interfaces/builtin: add support for content "source" section (v2)

### PR DESCRIPTION
**Pull request by zyga**
This branch alters the content interface to support the new syntax using
"source" section on content slots. This allows a content slot to share a
number of entries (both read and write at the same time).

When a "source" section is in use, the target directory on the plug
becomes a spool-like directory where content connections from various
connected slots is aggregated.

In addition, when entries would clash they are now automatically
renamed, which should help supporting plugin-type sharing.

This change itself is simple but will only work after the automatic hole poking
patches are merged (#4471 and #4472).

Signed-off-by: Zygmunt Krynicki zygmunt.krynicki@canonical.com
Forum: https://forum.snapcraft.io/t/improvements-in-the-content-interface/2387/4

### CODE REVIEW

File: interfaces/mount/spec.go [hunk 1]
```diff
@@ -43,6 +46,18 @@ func (spec *Specification) AddMountEntry(e Entry) error {
...
 	copy(result, spec.mountEntries)
+	// Number each entry, in case we get clashes this will automatically give
+	// them unique names.
+	count := make(map[string]int, len(spec.mountEntries))
+	for i := range result {
+		path := result[i].Dir
+		count[path] += 1
+		if c := count[path]; c > 1 {
+			newDir := fmt.Sprintf("%s-%d", result[i].Dir, c)
+			logger.Noticef("renaming mount entry for directory %q to %q to avoid a clash", result[i].Dir, newDir)
```
Comment (bboozzoo) [inline interfaces/mount/spec.go]: Thanks for the log.

File: interfaces/mount/spec.go [hunk 2]
```diff
@@ -43,6 +46,18 @@ func (spec *Specification) AddMountEntry(e Entry) error {
 func (spec *Specification) MountEntries() []Entry {
```
Comment (bboozzoo) [inline interfaces/mount/spec.go]: I'm grepping through the codebase, and MountEntries seems to be called only from tests. Shouldn't the declashing code be a part of AddMountEntry?

Comment (zyga) [reply interfaces/mount/spec.go]: I'll fix the backend to use the right thing. I may even move the declash logic to the backend directly.

File: interfaces/builtin/content.go [hunk 3]
```diff
@@ -78,6 +78,16 @@ func (iface *contentInterface) BeforePrepareSlot(slot *snap.SlotInfo) error {
 		slot.Attrs["content"] = slot.Name
 	}
 
+	// Error if "read" or "write" are present alongside "source".
+	if _, ok := slot.Attrs["source"]; ok {
```
Comment (stolowski) [inline interfaces/builtin/content.go]: You can use `Attr(name, value) error` on PlugInfo/SlotInfo now, would be better for consistency, and also we might want o hide Attrs map in near future.

Comment (zyga) [reply interfaces/builtin/content.go]: This is not plug/slot info but an attribute loaded from Attrs already

Comment (stolowski) [reply interfaces/builtin/content.go]: No, this **is** SlotInfo (and PlugInfo, respectively, in BeforePreparePlug below), so you can use Attrer interface over it via `if err := slot.Attr("source", &source); err {...`. Note that `path(...)` helper does exactly that and works both for Plug/SlotInfo and ConnectedPlug/Slot.

Comment (zyga) [reply interfaces/builtin/content.go]: Ah sorry, I misread the diff, you are totally right

Comment (zyga) [reply interfaces/builtin/content.go]: Done

File: interfaces/mount/spec.go (hunk 2 above)
Comment (zyga) [reply interfaces/mount/spec.go]: Fixed now, I will look at moving it to the backend

File: interfaces/builtin/content.go [hunk 4]
```diff
@@ -78,6 +78,18 @@ func (iface *contentInterface) BeforePrepareSlot(slot *snap.SlotInfo) error {
 		slot.Attrs["content"] = slot.Name
 	}
 
+	// Error if "read" or "write" are present alongside "source".
+	var unused map[string]interface{}
```
Comment (stolowski) [inline interfaces/builtin/content.go]: Note that this will only catch `source` if it's a map. But that's probably fine?

Comment (zyga) [reply interfaces/builtin/content.go]: Aha, good catch. I will change this to interface{} and ensure it is tested.

**Comment by jdstrand** [comment]
Can you please update https://forum.snapcraft.io/t/the-content-interface/1074?

**Comment by zyga** [comment]
@jdstrand that's a good idea. I will do that tomorrow.

**Pull request merged**